### PR TITLE
tests: improve RunTests.cmake

### PR DESCRIPTION
**Pull request by blueyed**
- allow for passing in BUSTED_ARGS via env
- quote values of TEST_TAG/TEST_FILTER

  Previously TEST_FILTER="'foo bar'" was required.

This allows for:

    make functionaltest TEST_FILE=test/functional/terminal/tui_spec.lua \
      BUSTED_ARGS="--no-keep-going --shuffle" \
      TEST_FILTER="TUI background color handles"

### CODE REVIEW

**Comment by blueyed** [comment]
JFI: this is the base for enabling parallel tests - which was inspired by recent tests adding more sleep, which made it obvious that running them in parallel is useful.

File: cmake/RunTests.cmake [hunk 1]
```diff
@@ -24,32 +24,37 @@ if(BUSTED_OUTPUT_TYPE STREQUAL junit)
...
     ${TEST_PATH}
   WORKING_DIRECTORY ${WORKING_DIR}
   ERROR_VARIABLE err
   RESULT_VARIABLE res
   ${EXTRA_ARGS})
 
-file(REMOVE ${WORKING_DIR}/Xtest_rplugin_manifest)
-file(REMOVE_RECURSE ${WORKING_DIR}/Xtest_xdg)
-file(REMOVE_RECURSE ${WORKING_DIR}/Xtest-tmpdir)
+file(REMOVE_RECURSE ${BUILD_DIR}/Xtest_*)
```
Comment (justinmk) [inline cmake/RunTests.cmake]: I assume you tested this but I could not find a reference about `file` command supporting wildcards. Generally cmake expects `GLOB` to be used then stored in a list which is then passed as an argument.

File: cmake/RunTests.cmake [hunk 2]
```diff
@@ -24,32 +24,37 @@ if(BUSTED_OUTPUT_TYPE STREQUAL junit)
...
+separate_arguments(BUSTED_ARGS)
+
 if(DEFINED ENV{TEST_TAG} AND NOT "$ENV{TEST_TAG}" STREQUAL "")
-  set(TEST_TAG "--tags=$ENV{TEST_TAG}")
+  list(APPEND BUSTED_ARGS "--tags='$ENV{TEST_TAG}'")
 endif()
 
 if(DEFINED ENV{TEST_FILTER} AND NOT "$ENV{TEST_FILTER}" STREQUAL "")
-  set(TEST_FILTER "--filter=$ENV{TEST_FILTER}")
+  list(APPEND BUSTED_ARGS "--filter='$ENV{TEST_FILTER}'")
```
Comment (justinmk) [inline cmake/RunTests.cmake]: Unless you are certain the single-quotes are needed I would leave them out. CMake generally adds the correct platform-specific quotations where needed, and adding quotes can cause surprising problems .

Comment (blueyed) [reply cmake/RunTests.cmake]: It is needed - see the description of https://github.com/neovim/neovim/pull/10239/commits/e119c9ce682feb829761c1870d8e47387baac454.

Comment (justinmk) [reply cmake/RunTests.cmake]: I would guess that single-quotes won't work on Windows if cmd.exe is involved somehow. Try this instead:

    list(APPEND BUSTED_ARGS --filter="$ENV{TEST_FILTER}")

**Comment by justinmk** [comment]
> JFI: this is the base for enabling parallel tests - which was inspired by recent tests adding more sleep, which made it obvious that running them in parallel is useful.

great idea, this will be extremely valuable. Hopefully there can still be a non-parallel fallback mode (or specify job number as integer) in case we run into problems.

**Comment by justinmk** [comment]
> JFI: this is the base for enabling parallel tests - which was inspired by recent tests adding more sleep, which made it obvious that running them in parallel is useful.

great idea, this will be extremely valuable. Hopefully there can still be a non-parallel fallback mode (or specify job number as integer) in case we run into problems.

BTW @ZyX-I  has related pending PR here: https://github.com/neovim/neovim/pull/8348

**Review by justinmk — APPROVED** [review]
see comments

File: cmake/RunTests.cmake (hunk 2 above)
Comment (blueyed) [reply cmake/RunTests.cmake]: This appears to work as well, will use that.

Comment (justinmk) [reply cmake/RunTests.cmake]: Actually, I bet it works without any quotes:

     list(APPEND BUSTED_ARGS --filter=$ENV{TEST_FILTER})

File: cmake/RunTests.cmake (hunk 1 above)
Comment (blueyed) [reply cmake/RunTests.cmake]: Ii thought so, but apparently you catched something there: fixed that.

File: cmake/RunTests.cmake (hunk 2 above)
Comment (blueyed) [reply cmake/RunTests.cmake]: Yeah, might be true given that is a list now.

File: cmake/RunTests.cmake [hunk 3]
```diff
@@ -16,40 +16,47 @@ endif()
...
   WORKING_DIRECTORY ${WORKING_DIR}
   ERROR_VARIABLE err
   RESULT_VARIABLE res
   ${EXTRA_ARGS})
 
-file(REMOVE ${WORKING_DIR}/Xtest_rplugin_manifest)
-file(REMOVE_RECURSE ${WORKING_DIR}/Xtest_xdg)
-file(REMOVE_RECURSE ${WORKING_DIR}/Xtest-tmpdir)
+file(GLOB RM_FILES ${BUILD_DIR}/Xtest_*)
+file(REMOVE ${RM_FILES})
```
Comment (justinmk) [inline cmake/RunTests.cmake]: Why not REMOVE_RECURSE?

Comment (blueyed) [reply cmake/RunTests.cmake]: Thought I've seen an error with it, but it appears to work.Will change it.

File: cmake/RunTests.cmake (hunk 2 above)
Comment (blueyed) [reply cmake/RunTests.cmake]: No.. check `TEST_FILE=test/functional/terminal/tui_spec.lua TEST_FILTER="TUI background color handles" make functionaltest`:

**Comment by blueyed** [comment]
Squashed.  @justinmk please review again.
As for running tests in parallel I'll postpone this for another PR, but having `BUSTED_ARGS` already is good also.

File: cmake/RunTests.cmake [hunk 4]
```diff
@@ -16,40 +16,47 @@ endif()
...
 if(BUSTED_OUTPUT_TYPE STREQUAL junit)
   set(EXTRA_ARGS OUTPUT_FILE ${BUILD_DIR}/${TEST_TYPE}test-junit.xml)
 endif()
 
+set(BUSTED_ARGS $ENV{BUSTED_ARGS})
+separate_arguments(BUSTED_ARGS)
+
 if(DEFINED ENV{TEST_TAG} AND NOT "$ENV{TEST_TAG}" STREQUAL "")
-  set(TEST_TAG "--tags=$ENV{TEST_TAG}")
+  list(APPEND BUSTED_ARGS "--tags='$ENV{TEST_TAG}'")
```
Comment (justinmk) [inline cmake/RunTests.cmake]: same as the other case, single-quotes should be avoided.

**Pull request merged**